src/MechaSwap.sol

```diff
@@ -0,0 +1,141 @@
+    event Init(address indexed user, uint256 amountX, uint256 amountY, uint256 amountZ);
+
+    /// @notice emitted when a user contributes liquidity to the pool
+    /// @param user address of user that is providing liquidity to the pool
+    /// @param amountX amount of token X that is being provided as liquidity
+    /// @param amountY amount of token Y that is being provided as liquidity 
+    /// @param amountZ amount of LP tokens that are being minted to the user
+    event LiquidityProvided(address indexed user, uint256 amountX, uint256 amountY, uint256 amountZ);
+
+    /// @notice emitted when a user removes liquidity from the pool
+    /// @param user address of user that is removing liquidity from the pool
+    /// @param amountX amount of token X that is being removed from liquidity
+    /// @param amountY amount of token Y that is being removed from liquidity
+    /// @param amountZ amount of LP tokens that are being burned from the user
+    event LiquidityRemoved(address indexed user, uint256 amountX, uint256 amountY, uint256 amountZ);
+
+    /// @notice emitted when a usere swaps tokens within the pool
+    /// @param user address of user that is swapping tokens
+    /// @param xIn amount of token X that is being swapped into the pool
+    /// @param yIn amount of token Y that is being swapped into the pool
+    /// @param xOut amount of token X that is being swapped out of the pool
+    /// @param yOut amount of token Y that is being swapped out of the pool
+    event Swap(address indexed user, uint256 xIn, uint256 yIn, uint256 xOut, uint256 yOut);
+
+    constructor(IERC20 tokenX_, IERC20 tokenY_) ERC20("MechaSwapLP", "MSLP", 18) {
+        tokenX = tokenX_;
+        tokenY = tokenY_;
+    }
+
+    /// @notice used to intiate a liquidity pool, can only be done once
+    /// @param amountX the amount of token X that'll be added to the pool 
+    /// @param amountY the amount of token Y that'll be added to the pool
+    /// @return z the amount of LP tokens that'll be minted to the user
+    function init(uint256 amountX, uint256 amountY) external returns(uint256 z){
+        require(x_0 == 0 && y_0 == 0, "MechaSwap:Pool already initiated");
+        x_0 += amountX;
+        y_0 += amountY;
+        z = amountX * amountY;
+        tokenX.safeTransferFrom(msg.sender, address(this), amountX);
+        tokenY.safeTransferFrom(msg.sender, address(this), amountY);
+        _mint(msg.sender, z);
+        emit Init(msg.sender, amountX, amountY, z);
+    }
+
+    /// @notice used to add liquidity to an existing pool
+    /// @param amountX the amount of token X that'll be added to the pool
+    /// @return z the amount of LP tokens that'll be minted to the user
+    function addLiquidity(uint256 amountX) external returns(uint256 z){
+        require(x_0 > 0, "MechaSwap:Pool not intiated");
+        uint256 amountY = amountX * y_0 / x_0;
+        x_0 += amountX;
+        y_0 += amountY;
+        z = amountX / x_0;
```

Comment:
This doesn't seem right. `z` will always be less than one, and rounded to zero.

In the requirement, `totalSupply` is mentioned. The user gets new tokens in the same proportion to `totalSupply`, as what he contributed to the reserves.